File: mobile_app/android/app/src/main/java/com/nwo/robotics/FunctionGemmaManager.kt
package com.nwo.robotics

import android.content.Context
import android.util.Log
import com.google.ai.edge.litert.Interpreter
import com.google.ai.edge.litert.GpuDelegate
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.io.FileInputStream

/**
 * FunctionGemma Manager for NWO Robotics
 * Handles on-device inference for voice-to-function calling
 */
class FunctionGemmaManager(private val context: Context) {
    
    companion object {
        private const val TAG = "FunctionGemmaManager"
        private const val MODEL_PATH = "functiongemma_270m.tflite"
        private const val MAX_TOKENS = 512
        private const val CONFIDENCE_THRESHOLD = 0.85
    }
    
    private var interpreter: Interpreter? = null
    private var gpuDelegate: GpuDelegate? = null
    private val functionSchemas: JSONObject
    private val commandCache = mutableMapOf<String, JSONObject>()
    
    // Function registry
    private val availableFunctions = listOf(
        "robot_command",
        "swarm_deploy", 
        "sensor_activate",
        "slam_start",
        "navigation_goto",
        "calibration_run",
        "emergency_stop",
        "status_check",
        "manipulator_control",
        "task_queue_submit",
        "patrol_route",
        "return_to_base",
        "follow_me",
        "voice_interaction"
    )
    
    init {
        functionSchemas = loadFunctionSchemas()
        initializeInterpreter()
    }
    
    /**
     * Initialize LiteRT interpreter with GPU acceleration
     */
    private fun initializeInterpreter() {
        try {
            val modelBuffer = loadModelFile()
            
            val options = Interpreter.Options().apply {
                // Try GPU first, fallback to CPU
                try {
                    gpuDelegate = GpuDelegate()
                    addDelegate(gpuDelegate)
                    Log.i(TAG, "GPU acceleration enabled")
                } catch (e: Exception) {
                    Log.w(TAG, "GPU not available, using CPU", e)
                    setNumThreads(4)
                }
                setUseXNNPACK(true)
            }
            
            interpreter = Interpreter(modelBuffer, options)
            Log.i(TAG, "FunctionGemma interpreter initialized")
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize interpreter", e)
            throw RuntimeException("Model initialization failed", e)
        }
    }
    
    /**
     * Load TFLite model from assets
     */
    private fun loadModelFile(): MappedByteBuffer {
        val fileDescriptor = context.assets.openFd(MODEL_PATH)
        val inputStream = FileInputStream(fileDescriptor.fileDescriptor)
        val fileChannel = inputStream.channel
        val startOffset = fileDescriptor.startOffset
        val declaredLength = fileDescriptor.declaredLength
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
    }
    
    /**
     * Load function schemas from JSON
     */
    private fun loadFunctionSchemas(): JSONObject {
        val jsonString = context.assets.open("nwo_robotics_functions.json")
            .bufferedReader()
            .use { it.readText() }
        return JSONObject(jsonString)
    }
    
    /**
     * Process voice/text command and return function calls
     */
    fun processCommand(
        command: String,
        contextData: Map<String, Any> = emptyMap()
    ): FunctionCallResult {
        
        // Check cache first
        commandCache[command]?.let {
            Log.d(TAG, "Cache hit for command: $command")
            return FunctionCallResult.Success(it, cached = true)
        }
        
        try {
            // Build prompt with system instructions and function schemas
            val prompt = buildPrompt(command, contextData)
            
            // Run inference
            val inputIds = tokenize(prompt)
            val outputIds = runInference(inputIds)
            val response = detokenize(outputIds)
            
            // Parse function calls from response
            val functionCalls = parseFunctionCalls(response)
            
            if (functionCalls.isEmpty()) {
                return FunctionCallResult.Error("No valid function calls generated")
            }
            
            // Validate and cache
            val validatedCalls = validateFunctionCalls(functionCalls)
            val result = JSONObject().apply {
                put("calls", validatedCalls)
                put("confidence", calculateConfidence(validatedCalls))
                put("original_command", command)
            }
            
            // Cache successful results
            if (result.getDouble("confidence") > CONFIDENCE_THRESHOLD) {
                commandCache[command] = result
            }
            
            return FunctionCallResult.Success(result, cached = false)
            
        } catch (e: Exception) {
            Log.e(TAG, "Command processing failed", e)
            return FunctionCallResult.Error(e.message ?: "Unknown error")
        }
    }
    
    /**
     * Build prompt with system instructions and schemas
     */
    private fun buildPrompt(command: String, context: Map<String, Any>): String {
        val systemPrompt = context.assets.open("system_prompt.txt")
            .bufferedReader()
            .use { it.readText() }
        
        val schemaSummary = buildSchemaSummary()
        
        return """
            $systemPrompt
            
            Available Functions:
            $schemaSummary
            
            User Command: "$command"
            
            Respond with JSON function calls:
            {"calls": [{"name": "function_name", "arguments": {...}}]}
        """.trimIndent()
    }
    
    /**
     * Build compact schema summary for prompt
     */
    private fun buildSchemaSummary(): String {
        val functions = functionSchemas.getJSONArray("functions")
        val summary = StringBuilder()
        
        for (i in 0 until functions.length()) {
            val func = functions.getJSONObject(i)
            summary.append("- ${func.getString("name")}: ${func.getString("description")}\n")
        }
        
        return summary.toString()
    }
    
    /**
     * Simple tokenization (replace with proper tokenizer)
     */
    private fun tokenize(text: String): IntArray {
        // TODO: Integrate SentencePiece tokenizer
        // Placeholder: convert to byte array and pad
        val tokens = mutableListOf<Int>()
        tokens.add(1) // BOS token
        text.toByteArray().forEach { byte ->
            tokens.add(byte.toInt() and 0xFF)
        }
        tokens.add(2) // EOS token
        return tokens.toIntArray()
    }
    
    /**
     * Run model inference
     */
    private fun runInference(inputIds: IntArray): IntArray {
        val interpreter = this.interpreter ?: throw IllegalStateException("Interpreter not initialized")
        
        // Prepare input/output buffers
        val inputBuffer = Array(1) { inputIds }
        val outputBuffer = Array(1) { IntArray(MAX_TOKENS) }
        
        // Run inference
        interpreter.run(inputBuffer, outputBuffer)
        
        return outputBuffer[0]
    }
    
    /**
     * Simple detokenization (replace with proper tokenizer)
     */
    private fun detokenize(tokenIds: IntArray): String {
        // TODO: Integrate SentencePiece detokenizer
        val bytes = tokenIds.filter { it > 2 }.map { it.toByte() }.toByteArray()
        return String(bytes)
    }
    
    /**
     * Parse JSON function calls from model output
     */
    private fun parseFunctionCalls(response: String): JSONArray {
        return try {
            val json = JSONObject(response)
            json.getJSONArray("calls")
        } catch (e: Exception) {
            // Try to extract JSON from markdown code blocks
            val jsonRegex = "```json\\s*(.*?)\\s*```".toRegex(RegexOption.DOT_MATCHES_ALL)
            val match = jsonRegex.find(response)
            if (match != null) {
                JSONObject(match.groupValues[1]).getJSONArray("calls")
            } else {
                JSONArray()
            }
        }
    }
    
    /**
     * Validate function calls against schemas
     */
    private fun validateFunctionCalls(calls: JSONArray): JSONArray {
        val validated = JSONArray()
        
        for (i in 0 until calls.length()) {
            val call = calls.getJSONObject(i)
            val funcName = call.getString("name")
            
            if (funcName !in availableFunctions) {
                Log.w(TAG, "Unknown function: $funcName")
                continue
            }
            
            // Get schema for this function
            val schema = findFunctionSchema(funcName)
            if (schema == null) {
                Log.w(TAG, "Schema not found for: $funcName")
                continue
            }
            
            // Validate required parameters
            val params = call.getJSONObject("arguments")
            val required = schema.getJSONObject("parameters")
                .optJSONArray("required") ?: JSONArray()
            
            val missing = mutableListOf<String>()
            for (j in 0 until required.length()) {
                val req = required.getString(j)
                if (!params.has(req)) {
                    missing.add(req)
                }
            }
            
            if (missing.isNotEmpty()) {
                Log.w(TAG, "Missing required params for $funcName: $missing")
                // Fill defaults if available
                fillDefaultValues(params, schema.getJSONObject("parameters"))
            }
            
            validated.put(call)
        }
        
        return validated
    }
    
    /**
     * Find function schema by name
     */
    private fun findFunctionSchema(name: String): JSONObject? {
        val functions = functionSchemas.getJSONArray("functions")
        for (i in 0 until functions.length()) {
            val func = functions.getJSONObject(i)
            if (func.getString("name") == name) {
                return func
            }
        }
        return null
    }
    
    /**
     * Fill default values for missing parameters
     */
    private fun fillDefaultValues(params: JSONObject, schema: JSONObject) {
        val properties = schema.optJSONObject("properties") ?: return
        
        properties.keys().forEach { key ->
            if (!params.has(key)) {
                val prop = properties.getJSONObject(key)
                val default = prop.opt("default")
                if (default != null) {
                    params.put(key, default)
                }
            }
        }
    }
    
    /**
     * Calculate confidence score for function calls
     */
    private fun calculateConfidence(calls: JSONArray): Double {
        // Simple heuristic: more calls = lower confidence, validate params
        var score = 1.0
        
        if (calls.length() > 3) {
            score -= 0.1 * (calls.length() - 3)
        }
        
        for (i in 0 until calls.length()) {
            val call = calls.getJSONObject(i)
            if (call.getString("name") == "emergency_stop") {
                score += 0.05 // Boost for safety commands
            }
        }
        
        return score.coerceIn(0.0, 1.0)
    }
    
    /**
     * Check if command requires confirmation
     */
    fun requiresConfirmation(functionName: String): Boolean {
        return functionName in listOf("emergency_stop", "calibration_run", "swarm_deploy")
    }
    
    /**
     * Clear command cache
     */
    fun clearCache() {
        commandCache.clear()
        Log.i(TAG, "Command cache cleared")
    }
    
    /**
     * Release resources
     */
    fun release() {
        interpreter?.close()
        gpuDelegate?.close()
        interpreter = null
        gpuDelegate = null
        Log.i(TAG, "Resources released")
    }
    
    sealed class FunctionCallResult {
        data class Success(val data: JSONObject, val cached: Boolean) : FunctionCallResult()
        data class Error(val message: String) : FunctionCallResult()
    }
}

File: mobile_app/ios/NwoRoboticsController/FunctionGemmaManager.swift
import Foundation
import TensorFlowLite
import Accelerate

/// FunctionGemma Manager for NWO Robotics (iOS)
/// Handles on-device inference for voice-to-function calling
class FunctionGemmaManager {
    
    // MARK: - Constants
    private let modelName = "functiongemma_270m.tflite"
    private let maxTokens = 512
    private let confidenceThreshold = 0.85
    
    // MARK: - Properties
    private var interpreter: Interpreter?
    private let functionSchemas: [String: Any]
    private var commandCache: [String: [String: Any]] = [:]
    
    let availableFunctions = [
        "robot_command", "swarm_deploy", "sensor_activate", "slam_start",
        "navigation_goto", "calibration_run", "emergency_stop", "status_check",
        "manipulator_control", "task_queue_submit", "patrol_route",
        "return_to_base", "follow_me", "voice_interaction"
    ]
    
    // MARK: - Initialization
    init() throws {
        self.functionSchemas = try FunctionGemmaManager.loadFunctionSchemas()
        try initializeInterpreter()
    }
    
    deinit {
        interpreter = nil
    }
    
    // MARK: - Model Setup
    private func initializeInterpreter() throws {
        guard let modelPath = Bundle.main.path(forResource: "functiongemma_270m", ofType: "tflite") else {
            throw FunctionGemmaError.modelNotFound
        }
        
        var options = Interpreter.Options()
        options.threadCount = 4
        
        // Configure delegates
        var delegates: [Delegate] = []
        
        // Try Metal delegate for GPU acceleration
        if let metalDelegate = try? MetalDelegate() {
            delegates.append(metalDelegate)
            print("✓ Metal GPU acceleration enabled")
        } else {
            print("⚠ Metal not available, using CPU")
        }
        
        interpreter = try Interpreter(modelPath: modelPath, options: options, delegates: delegates)
        try interpreter?.allocateTensors()
        
        print("✓ FunctionGemma interpreter initialized")
    }
    
    private static func loadFunctionSchemas() throws -> [String: Any] {
        guard let url = Bundle.main.url(forResource: "nwo_robotics_functions", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw FunctionGemmaError.schemaLoadFailed
        }
        return json
    }
    
    // MARK: - Command Processing
    func processCommand(
        _ command: String,
        context: [String: Any] = [:]
    ) -> FunctionCallResult {
        
        // Check cache
        if let cached = commandCache[command] {
            print("📦 Cache hit for: \(command)")
            return .success(cached, cached: true)
        }
        
        do {
            // Build prompt
            let prompt = buildPrompt(command: command, context: context)
            
            // Tokenize and run inference
            let inputIds = tokenize(prompt)
            let outputIds = try runInference(inputIds: inputIds)
            let response = detokenize(outputIds)
            
            // Parse function calls
            guard let functionCalls = parseFunctionCalls(response),
                  !functionCalls.isEmpty else {
                return .error("No valid function calls generated")
            }
            
            // Validate
            let validatedCalls = validateFunctionCalls(functionCalls)
            let result: [String: Any] = [
                "calls": validatedCalls,
                "confidence": calculateConfidence(validatedCalls),
                "original_command": command
            ]
            
            // Cache if confident
            if let confidence = result["confidence"] as? Double,
               confidence > confidenceThreshold {
                commandCache[command] = result
            }
            
            return .success(result, cached: false)
            
        } catch {
            return .error(error.localizedDescription)
        }
    }
    
    // MARK: - Prompt Building
    private func buildPrompt(command: String, context: [String: Any]) -> String {
        let systemPrompt = loadSystemPrompt()
        let schemaSummary = buildSchemaSummary()
        
        return """
        \(systemPrompt)
        
        Available Functions:
        \(schemaSummary)
        
        User Command: "\(command)"
        
        Respond with JSON function calls:
        {"calls": [{"name": "function_name", "arguments": {...}}]}
        """
    }
    
    private func loadSystemPrompt() -> String {
        guard let url = Bundle.main.url(forResource: "system_prompt", withExtension: "txt"),
              let content = try? String(contentsOf: url) else {
            return "You are a robot control assistant. Convert natural language to function calls."
        }
        return content
    }
    
    private func buildSchemaSummary() -> String {
        guard let functions = functionSchemas["functions"] as? [[String: Any]] else {
            return ""
        }
        
        return functions.compactMap { funcDef -> String? in
            guard let name = funcDef["name"] as? String,
                  let desc = funcDef["description"] as? String else { return nil }
            return "- \(name): \(desc)"
        }.joined(separator: "\n")
    }
    
    // MARK: - Tokenization (Placeholder - integrate SentencePiece)
    private func tokenize(_ text: String) -> [Int32] {
        // TODO: Replace with actual SentencePiece tokenizer
        var tokens: [Int32] = [1] // BOS
        for byte in text.utf8 {
            tokens.append(Int32(byte))
        }
        tokens.append(2) // EOS
        return tokens
    }
    
    private func detokenize(_ tokens: [Int32]) -> String {
        // TODO: Replace with actual SentencePiece detokenizer
        let bytes = tokens.filter { $0 > 2 }.map { UInt8($0) }
        return String(bytes: bytes, encoding: .utf8) ?? ""
    }
    
    // MARK: - Inference
    private func runInference(inputIds: [Int32]) throws -> [Int32] {
        guard let interpreter = interpreter else {
            throw FunctionGemmaError.interpreterNotInitialized
        }
        
        // Prepare input tensor
        let inputShape = interpreter.input(at: 0).shape
        let inputData = Data(bytes: inputIds, count: inputIds.count * MemoryLayout<Int32>.size)
        
        try interpreter.copy(inputData, toInputAt: 0)
        try interpreter.invoke()
        
        // Get output
        let outputTensor = try interpreter.output(at: 0)
        let outputData = outputTensor.data
        
        // Convert to Int32 array
        var outputIds = [Int32](repeating: 0, count: outputData.count / MemoryLayout<Int32>.size)
        outputData.copyBytes(to: &outputIds, count: outputData.count)
        
        return Array(outputIds.prefix(maxTokens))
    }
    
    // MARK: - Parsing & Validation
    private func parseFunctionCalls(_ response: String) -> [[String: Any]]? {
        // Try direct JSON parsing
        if let data = response.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let calls = json["calls"] as? [[String: Any]] {
            return calls
        }
        
        // Try extracting from markdown code blocks
        let pattern = "```json\\s*(.*?)\\s*```"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .dotMatchesLineSeparators),
              let match = regex.firstMatch(in: response, range: NSRange(response.startIndex..., in: response)) else {
            return nil
        }
        
        if let range = Range(match.range(at: 1), in: response) {
            let jsonStr = String(response[range])
            if let data = jsonStr.data(using: .utf8),
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let calls = json["calls"] as? [[String: Any]] {
                return calls
            }
        }
        
        return nil
    }
    
    private func validateFunctionCalls(_ calls: [[String: Any]]) -> [[String: Any]] {
        return calls.compactMap { call -> [String: Any]? in
            guard let funcName = call["name"] as? String,
                  availableFunctions.contains(funcName) else {
                print("⚠ Unknown function: \(call["name"] ?? "nil")")
                return nil
            }
            
            // Find schema
            guard let functions = functionSchemas["functions"] as? [[String: Any]],
                  let schema = functions.first(where: { $0["name"] as? String == funcName }),
                  let params = schema["parameters"] as? [String: Any] else {
                return call
            }
            
            // Validate and fill defaults
            var validatedCall = call
            if var arguments = call["arguments"] as? [String: Any] {
                arguments = fillDefaultValues(arguments: arguments, schema: params)
                validatedCall["arguments"] = arguments
            }
            
            return validatedCall
        }
    }
    
    private func fillDefaultValues(arguments: [String: Any], schema: [String: Any]) -> [String: Any] {
        var result = arguments
        
        guard let properties = schema["properties"] as? [String: [String: Any]] else {
            return result
        }
        
        for (key, prop) in properties {
            if result[key] == nil, let defaultValue = prop["default"] {
                result[key] = defaultValue
            }
        }
        
        return result
    }
    
    private func calculateConfidence(_ calls: [[String: Any]]) -> Double {
        var score = 1.0
        
        // Penalty for many calls
        if calls.count > 3 {
            score -= 0.1 * Double(calls.count - 3)
        }
        
        // Boost for safety commands
        for call in calls {
            if call["name"] as? String == "emergency_stop" {
                score += 0.05
            }
        }
        
        return min(max(score, 0.0), 1.0)
    }
    
    // MARK: - Utilities
    func requiresConfirmation(_ functionName: String) -> Bool {
        return ["emergency_stop", "calibration_run", "swarm_deploy"].contains(functionName)
    }
    
    func clearCache() {
        commandCache.removeAll()
        print("🗑 Cache cleared")
    }
    
    // MARK: - Types
    enum FunctionCallResult {
        case success([String: Any], cached: Bool)
        case error(String)
    }
    
    enum FunctionGemmaError: Error {
        case modelNotFound
        case schemaLoadFailed
        case interpreterNotInitialized
        case inferenceFailed
    }
}
